File: tests/math/swap_math.ts
import Decimal from "decimal.js";
import {
  calculate_curve,
  calculate_market_cap,
  calculate_price,
  ceil_div,
  CURVE_LAST_PARAMS,
  CURVES,
  find_root,
  MAX_TOKEN_SUPPLY,
  search_curve,
} from "./token_math";

/**
 * Computes the amount of token to buy or sell with fee.
 * @param amount Amount of token to buy or sell.
 * @param remaining_supply Remaining amount of token in the bounding curve.
 * @param buy True if buying token, false if selling token.
 */
export function compute_swap_with_fee(
  amount: bigint | number,
  remaining_supply: bigint | number,
  buy: boolean
): ComputeSwapOutput {
  let y = compute_swap(BigInt(amount), BigInt(remaining_supply), buy);
  let fee = compute_fee(y);
  let total: bigint;
  if (buy) {
    total = y + fee;
  } else {
    total = y - fee;
  }
  return { y, fee, total };
}

export function compute_swap(
  amount: bigint,
  remaining_token_supply: bigint,
  buy: boolean
): bigint {
  let target_token_supply: bigint;
  if (buy) {
    target_token_supply = remaining_token_supply - amount;
    if (target_token_supply == BigInt(0)) {
      throw new Error("Cannot buy all remaining supply");
    }
  } else {
    target_token_supply = remaining_token_supply + amount;
    if (target_token_supply > MAX_TOKEN_SUPPLY) {
      throw new Error("Cannot sell more than max token supply");
    }
  }

  let native_amount = BigInt(0);
  if (buy) {
    let start_native_amount = BigInt(0);
    let start_native_amount_not_set = true;
    let end_native_amount = BigInt(0);
    for (let curve of CURVES) {
      if (remaining_token_supply > curve.token_supply_at_boundary) {
        if (start_native_amount_not_set) {
          start_native_amount_not_set = false;
          start_native_amount = calculate_curve(
            remaining_token_supply,
            false,
            curve
          );
        }

        if (target_token_supply >= curve.token_supply_at_boundary) {
          end_native_amount = calculate_curve(target_token_supply, true, curve);
          break;
        } else {
          // reach the curve boundary
          remaining_token_supply = curve.token_supply_at_boundary;
        }
      }
    }
    native_amount = end_native_amount - start_native_amount;
  } else {
    let end_native_amount = BigInt(0);
    let start_native_amount = BigInt(0);
    for (let i = CURVES.length - 1; i >= 0; i--) {
      let curve = CURVES[i];
      let token_supply_at_start_boundary: bigint;
      if (i == 0) {
        token_supply_at_start_boundary = MAX_TOKEN_SUPPLY;
      } else {
        token_supply_at_start_boundary = CURVES[i - 1].token_supply_at_boundary;
      }

      if (remaining_token_supply < token_supply_at_start_boundary) {
        if (end_native_amount == BigInt(0)) {
          end_native_amount = calculate_curve(
            remaining_token_supply,
            false,
            curve
          );
        }

        if (target_token_supply <= token_supply_at_start_boundary) {
          start_native_amount = calculate_curve(
            target_token_supply,
            true,
            curve
          );
          break;
        } else {
          // go back the previous curve boundary
          remaining_token_supply = token_supply_at_start_boundary;
        }
      }
    }
    if (end_native_amount > start_native_amount) {
      native_amount = end_native_amount - start_native_amount;
    }
  }

  return native_amount;
}

export function compute_buy_token_exact_in_with_fee(
  pay_amount: bigint | number,
  remaining_token_supply: bigint | number
): ComputeBuyExactInOutput {
  let buy_amount = compute_buy_token_exact_in(
    BigInt(pay_amount),
    BigInt(remaining_token_supply)
  );
  let fee = compute_fee(pay_amount);
  return {
    buy_amount: buy_amount,
    y: BigInt(pay_amount),
    fee: fee,
    total: BigInt(pay_amount) + fee,
  };
}

export function compute_buy_token_exact_in(
  pay_amount: bigint,
  remaining_token_supply: bigint
): bigint {
  let start_native_amount = BigInt(0);
  let start_native_amount_not_set = true;
  let buy_amount = BigInt(0);
  for (let curve of CURVES) {
    if (remaining_token_supply > curve.token_supply_at_boundary) {
      if (start_native_amount_not_set) {
        start_native_amount_not_set = false;
        start_native_amount = calculate_curve(
          remaining_token_supply,
          false,
          curve
        );
      }

      let remaining_native_amount =
        curve.native_amount_at_boundary - start_native_amount;
      if (
        pay_amount < remaining_native_amount ||
        curve.k_with_multiplier_sol == CURVE_LAST_PARAMS.k_with_multiplier_sol
      ) {
        // still in the curve
        start_native_amount = calculate_curve(
          remaining_token_supply,
          true,
          curve
        );
        buy_amount += find_root(
          remaining_token_supply,
          start_native_amount,
          pay_amount,
          curve
        );
        break;
      } else if (pay_amount == remaining_native_amount) {
        // reach the curve boundary and the remaining native amount is exactly the same as the boundary
        buy_amount += remaining_token_supply - curve.token_supply_at_boundary;
        break;
      } else {
        // reach the curve boundary
        buy_amount += remaining_token_supply - curve.token_supply_at_boundary;
        remaining_token_supply = curve.token_supply_at_boundary;
        pay_amount -= remaining_native_amount;
        start_native_amount = curve.native_amount_at_boundary;
      }
    }
  }
  return buy_amount;
}

export function compute_fee(amount: bigint | number): bigint {
  return ceil_div(BigInt(amount), BigInt(100)); // 1%
}

export function curve_points(
  remaining_token_supply: bigint | number,
  max_points: number = 1000,
  x_axis_max_threshold: number = 300, // 3x
  x_axis_min_y: bigint | number = BigInt(200) * BigInt(1e9) // 200 SOL
): Array<CurvePoint> {
  if (x_axis_max_threshold < 100) {
    throw new Error("Invalid max threshold");
  }
  if (max_points < 2) {
    throw new Error("Invalid max points");
  }

  let points: Array<CurvePoint> = [];

  let current_sold = MAX_TOKEN_SUPPLY - BigInt(remaining_token_supply);
  let current_y = compute_swap(current_sold, MAX_TOKEN_SUPPLY, true);
  let max_threshold_y =
    (current_y * BigInt(x_axis_max_threshold)) / BigInt(100);
  let max_y = compute_swap(
    MAX_TOKEN_SUPPLY - BigInt(1),
    MAX_TOKEN_SUPPLY,
    true
  );
  if (max_threshold_y > max_y) {
    max_threshold_y = max_y;
  }
  if (max_threshold_y < BigInt(x_axis_min_y)) {
    max_threshold_y = BigInt(x_axis_min_y);
  }

  let max_buy_amount = compute_buy_token_exact_in(
    max_threshold_y,
    MAX_TOKEN_SUPPLY
  );

  let min_buy_amount = BigInt(0);

  let step = ceil_div(max_buy_amount - min_buy_amount, BigInt(max_points - 1));

  let first_current_point: undefined | number;
  for (let i = 0; i < max_points; i++) {
    let buy_amount = min_buy_amount + step * BigInt(i);
    let y = BigInt(0);
    if (buy_amount > BigInt(0)) {
      y = compute_swap(buy_amount, MAX_TOKEN_SUPPLY, true);
    }

    let point = {
      buy_amount: buy_amount,
      y: y,
      current: false,
      market_cap: calculate_market_cap(
        calculate_price(
          MAX_TOKEN_SUPPLY - buy_amount,
          y,
          search_curve(MAX_TOKEN_SUPPLY - buy_amount)
        )
      ),
    };
    points.push(point);

    if (
      first_current_point === undefined &&
      buy_amount >= BigInt(current_sold)
    ) {
      first_current_point = Number(i);
    }
  }

  let point = points[first_current_point];
  if (point.y == current_sold) {
    point.current = true;
  } else {
    // replace the first point with the current point
    let y = compute_swap(current_sold, MAX_TOKEN_SUPPLY, true);
    point.buy_amount = current_sold;
    point.y = y;
    point.current = true;
    point.market_cap = calculate_market_cap(
      calculate_price(
        BigInt(remaining_token_supply),
        y,
        search_curve(BigInt(remaining_token_supply))
      )
    );
  }
  return points;
}

export type ComputeSwapOutput = {
  /// Amount of token(SOL) to pay or receive
  y: bigint;
  /// Amount of fee to pay
  fee: bigint;
  /// Total amount of token(SOL) to pay or receive
  total: bigint;
};

export type ComputeBuyExactInOutput = {
  /// Amount of token to buy
  buy_amount: bigint;
  /// Amount of token(SOL) to pay
  y: bigint;
  /// Amount of fee to pay
  fee: bigint;
  /// Total amount of token(SOL) to pay
  total: bigint;
};

export type CurvePoint = {
  /// Amount of token to buy
  buy_amount: bigint;
  /// Amount of token(SOL) to pay
  y: bigint;
  /// Market cap
  market_cap: Decimal;
  /// True if the point is the current point
  current: boolean;
};

File: programs/burn/src/math/swap_math.rs
use crate::{constants::MAX_TOKEN_SUPPLY, errors::Error as MyError};
use anchor_lang::error::Error;

use super::{math::ceil_div, token_math};

/// Computes the native amount to be paid when buying or received when selling.
///
/// # Parameters
/// - `amount`: The amount of token to be swapped.
/// - `remaining_supply`: The remaining token supply on the bounding curve.
/// - `buy`: If true, buying, if false, selling.
///
/// <div class="warning">
///
/// The caller MUST ensure that the `amount` does not exceed `remaining_supply` when buying
/// and `constants::MAX_TOKEN_SUPPLY - remaining_supply` when selling.
///
/// </div>
pub fn compute_swap(amount: u64, remaining_token_supply: u64, buy: bool) -> Result<u64, Error> {
    let target_token_supply = if buy {
        remaining_token_supply - amount
    } else {
        remaining_token_supply + amount
    };

    let mut remaining_token_supply = remaining_token_supply;
    let mut native_amount: u128 = 0;
    if buy {
        let mut start_native_amount: u128 = 0;
        let mut start_native_amount_not_set: bool = true;
        let mut end_native_amount: u128 = 0;
        for curve in token_math::CURVES.iter() {
            if remaining_token_supply > curve.token_supply_at_boundary {
                if start_native_amount_not_set {
                    start_native_amount_not_set = false;
                    start_native_amount = token_math::calculate_curve(remaining_token_supply, false, curve);
                }

                if target_token_supply >= curve.token_supply_at_boundary {
                    end_native_amount = token_math::calculate_curve(target_token_supply, true, curve);
                    break;
                } else {
                    // reach the curve boundary
                    remaining_token_supply = curve.token_supply_at_boundary;
                }
            }
        }
        native_amount = end_native_amount - start_native_amount;
    } else {
        let mut i: usize = token_math::CURVES.len() - 1;
        let mut end_native_amount: u128 = 0;
        let start_native_amount: u128;
        loop {
            let curve = token_math::CURVES[i];
            let token_supply_at_start_boundary: u64;
            if i == 0 {
                token_supply_at_start_boundary = MAX_TOKEN_SUPPLY;
            } else {
                token_supply_at_start_boundary = token_math::CURVES[i - 1].token_supply_at_boundary;
            }

            if remaining_token_supply < token_supply_at_start_boundary {
                if end_native_amount == 0 {
                    end_native_amount = token_math::calculate_curve(remaining_token_supply, false, curve);
                }

                if target_token_supply <= token_supply_at_start_boundary {
                    start_native_amount = token_math::calculate_curve(target_token_supply, true, curve);
                    break;
                } else {
                    // go back the previous curve boundary
                    remaining_token_supply = token_supply_at_start_boundary;
                }
            }

            debug_assert!(i >= 1);
            i -= 1;
        }
        if end_native_amount > start_native_amount {
            native_amount = end_native_amount - start_native_amount;
        }
    }

    if native_amount > u64::MAX as u128 {
        return Err(MyError::TooMuchNativeTokenRequired.into());
    }

    Ok(native_amount as u64)
}

/// Computes the amount of token to be bought with the given pay amount.
pub fn compute_buy_token_exact_in(pay_amount: u64, remaining_token_supply: u64) -> Result<u64, Error> {
    let mut start_native_amount: u128 = 0;
    let mut start_native_amount_not_set: bool = true;
    let mut pay_amount = pay_amount as u128;
    let mut remaining_token_supply = remaining_token_supply;
    let mut buy_amount = 0;
    for curve in token_math::CURVES.iter() {
        if remaining_token_supply > curve.token_supply_at_boundary {
            if start_native_amount_not_set {
                start_native_amount_not_set = false;
                start_native_amount = token_math::calculate_curve(remaining_token_supply, false, curve);
            }

            let remaining_native_amount: u128 = curve.native_amount_at_boundary as u128 - start_native_amount;
            if pay_amount < remaining_native_amount
                || curve.k_with_multiplier_sol == token_math::CURVE_LAST_PARAMS.k_with_multiplier_sol
            {
                // still in the curve
                start_native_amount = token_math::calculate_curve(remaining_token_supply, true, curve); // use rounding up to ensure that the number of tokens bought is less
                buy_amount += token_math::find_root(remaining_token_supply, start_native_amount, pay_amount, curve)?;
                break;
            } else if pay_amount == remaining_native_amount {
                // reach the curve boundary and the remaining native amount is exactly the same as the boundary
                buy_amount += remaining_token_supply - curve.token_supply_at_boundary;
                break;
            } else {
                // reach the curve boundary
                buy_amount += remaining_token_supply - curve.token_supply_at_boundary;
                remaining_token_supply = curve.token_supply_at_boundary;
                pay_amount -= remaining_native_amount;
                start_native_amount = curve.native_amount_at_boundary as u128;
            }
        }
    }
    Ok(buy_amount)
}

/// Splits an amount into buy amount and fee amount.
pub fn split_pay_amount(max_pay_amount: u64) -> Result<(u64, u64), Error> {
    // x * (1 + fee) = max_pay_amount
    // x = max_pay_amount / (1 + fee)
    // x = max_pay_amount / (1 + 1 / 100)
    // x = max_pay_amount * 100 / 101
    let pay_amount = ceil_div(max_pay_amount as u128 * 100, 101) as u64;
    let fee_amount = max_pay_amount - pay_amount;
    Ok((pay_amount, fee_amount))
}

/// Computes the fee amount.
pub fn compute_fee(amount: u64) -> u64 {
    ceil_div(amount as u128, 100) as u64 // 1%
}
#[cfg(test)]
mod tests {
    use anchor_lang::solana_program::native_token::LAMPORTS_PER_SOL;

    use super::*;
    use crate::constants::MAX_TOKEN_SUPPLY;

    #[test]
    #[should_panic(expected = "TooMuchNativeTokenRequired")]
    fn test_compute_swap_too_much_native_token_required() {
        compute_swap(1, 2, true).unwrap();
    }

    #[test]
    fn test_compute_swap_with_buy_and_curve_1() {
        let y = compute_swap(1, MAX_TOKEN_SUPPLY, true).unwrap();
        assert_eq!(y, 1);

        let y = compute_swap(1, token_math::CURVE_1_PARAMS.token_supply_at_boundary + 1, true).unwrap();
        assert_eq!(y, 1);
        let y = compute_swap(1, token_math::CURVE_1_PARAMS.token_supply_at_boundary, false).unwrap();
        assert_eq!(y, 0);

        let y = compute_swap(MAX_TOKEN_SUPPLY * 20 / 100, MAX_TOKEN_SUPPLY, true).unwrap();
        assert_eq!(y, token_math::CURVE_1_PARAMS.native_amount_at_boundary);

        let y = compute_swap(
            MAX_TOKEN_SUPPLY * 20 / 100,
            token_math::CURVE_1_PARAMS.token_supply_at_boundary,
            false,
        )
        .unwrap();
        assert_eq!(y, token_math::CURVE_1_PARAMS.native_amount_at_boundary);
    }

    #[test]
    fn test_compute_swap_with_buy_and_curve_2() {
        let y = compute_swap(1, token_math::CURVE_2_PARAMS.token_supply_at_boundary, true).unwrap();
        assert_eq!(y, 0 + 1);

        let y = compute_swap(1, token_math::CURVE_2_PARAMS.token_supply_at_boundary - 1, false).unwrap();
        assert_eq!(y, 0);

        let y = compute_swap(
            token_math::CURVE_1_PARAMS.token_supply_at_boundary - token_math::CURVE_2_PARAMS.token_supply_at_boundary,
            token_math::CURVE_1_PARAMS.token_supply_at_boundary,
            true,
        )
        .unwrap();
        assert_eq!(
            y,
            token_math::CURVE_2_PARAMS.native_amount_at_boundary - token_math::CURVE_1_PARAMS.native_amount_at_boundary
        );

        let y = compute_swap(
            token_math::CURVE_1_PARAMS.token_supply_at_boundary - token_math::CURVE_2_PARAMS.token_supply_at_boundary,
            token_math::CURVE_2_PARAMS.token_supply_at_boundary,
            false,
        )
        .unwrap();
        assert_eq!(
            y,
            token_math::CURVE_2_PARAMS.native_amount_at_boundary - token_math::CURVE_1_PARAMS.native_amount_at_boundary
        );
    }

    #[test]
    fn test_compute_swap_with_buy_and_curve_3() {
        let y = compute_swap(1, token_math::CURVE_2_PARAMS.token_supply_at_boundary, true).unwrap();
        assert_eq!(y, 0 + 1);
        let y = compute_swap(1, token_math::CURVE_2_PARAMS.token_supply_at_boundary - 1, false).unwrap();
        assert_eq!(y, 0);

        let y = compute_swap(1, MAX_TOKEN_SUPPLY * 5 / 100, true).unwrap();
        assert_eq!(y, 0 + 1);
        let y = compute_swap(1, MAX_TOKEN_SUPPLY * 5 / 100 - 1, false).unwrap();
        assert_eq!(y, 0);

        let y = compute_swap(1, MAX_TOKEN_SUPPLY * 3 / 100, true).unwrap();
        assert_eq!(y, 1 + 1);
        let y = compute_swap(1, MAX_TOKEN_SUPPLY * 3 / 100 - 1, false).unwrap();
        assert_eq!(y, 0);
    }

    #[test]
    fn test_compute_swap_with_buy_and_both_curve() {
        let pay = compute_swap(1, MAX_TOKEN_SUPPLY, true).unwrap();
        assert_eq!(pay, 0 + 1);
        let receive = compute_swap(1, MAX_TOKEN_SUPPLY - 1, false).unwrap();
        assert_eq!(receive, 0);

        let pay = compute_swap(1, token_math::CURVE_1_PARAMS.token_supply_at_boundary + 1, true).unwrap();
        assert_eq!(pay, 1);
        let receive = compute_swap(1, token_math::CURVE_1_PARAMS.token_supply_at_boundary, false).unwrap();
        assert_eq!(receive, 0);

        let pay = compute_swap(2, token_math::CURVE_1_PARAMS.token_supply_at_boundary + 1, true).unwrap();
        assert_eq!(pay, 1 + 1);
        let receive = compute_swap(2, token_math::CURVE_1_PARAMS.token_supply_at_boundary - 1, false).unwrap();
        assert_eq!(receive, 0);

        let pay = compute_swap(MAX_TOKEN_SUPPLY * 10 / 100 + 1, MAX_TOKEN_SUPPLY, true).unwrap();
        assert_eq!(pay, 3669105319 + 1);
        let receive = compute_swap(
            MAX_TOKEN_SUPPLY * 10 / 100 + 1,
            MAX_TOKEN_SUPPLY - 1 - MAX_TOKEN_SUPPLY * 10 / 100,
            false,
        )
        .unwrap();
        assert_eq!(receive, 3669105319);

        let pay = compute_swap(MAX_TOKEN_SUPPLY * 10 / 100, MAX_TOKEN_SUPPLY - 1, true).unwrap();
        assert_eq!(pay, 3669105319 + 1);
        let receive = compute_swap(
            MAX_TOKEN_SUPPLY * 10 / 100,
            MAX_TOKEN_SUPPLY - 1 - MAX_TOKEN_SUPPLY * 10 / 100,
            false,
        )
        .unwrap();
        assert_eq!(receive, 3669105318);

        let pay = compute_swap(MAX_TOKEN_SUPPLY * 95 / 100, MAX_TOKEN_SUPPLY, true).unwrap();
        assert_eq!(pay, 8725910156250);
        let receive = compute_swap(MAX_TOKEN_SUPPLY * 95 / 100, MAX_TOKEN_SUPPLY * 5 / 100, false).unwrap();
        assert_eq!(receive, 8725910156250);

        let pay = compute_swap(MAX_TOKEN_SUPPLY * 95 / 100 + 1, MAX_TOKEN_SUPPLY - 1, true).unwrap();
        assert_eq!(pay, 8725910156250 + 1);
        let receive = compute_swap(MAX_TOKEN_SUPPLY * 95 / 100 + 1, MAX_TOKEN_SUPPLY * 5 / 100 - 2, false).unwrap();
        assert_eq!(receive, 8725910156249);
    }

    #[test]
    fn test_compute_swap_with_buy_and_sell_on_curve_1() {
        let sell_y = compute_swap(MAX_TOKEN_SUPPLY * 1 / 10, MAX_TOKEN_SUPPLY * 7 / 10, false).unwrap();
        let buy_y = compute_swap(MAX_TOKEN_SUPPLY * 1 / 10, MAX_TOKEN_SUPPLY * 8 / 10, true).unwrap();
        assert!(buy_y > sell_y);
    }

    #[test]
    fn test_compute_swap_with_buy_and_sell_on_curve_2() {
        let sell_y = compute_swap(MAX_TOKEN_SUPPLY * 1 / 10, MAX_TOKEN_SUPPLY * 2 / 10, false).unwrap();
        let buy_y = compute_swap(MAX_TOKEN_SUPPLY * 1 / 10, MAX_TOKEN_SUPPLY * 3 / 10, true).unwrap();
        assert!(buy_y > sell_y);
    }

    #[test]
    fn test_compute_swap_with_buy_and_sell_on_both_curve() {
        let sell_y = compute_swap(MAX_TOKEN_SUPPLY * 2 / 10, MAX_TOKEN_SUPPLY * 4 / 10, false).unwrap();
        let buy_y = compute_swap(MAX_TOKEN_SUPPLY * 2 / 10, MAX_TOKEN_SUPPLY * 6 / 10, true).unwrap();
        assert!(buy_y > sell_y);
    }

    #[derive(Debug)]
    struct TestCase {
        remaining_supply: u64,
        actual_pay: u64,
    }

    #[test]
    fn test_compute_buy_token_exact_in() {
        let cases = [
            TestCase {
                remaining_supply: 10e8 as u64 * 1e6 as u64,
                actual_pay: 32767 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 10e8 as u64 * 1e6 as u64 - 1,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 999999999 * 1e6 as u64,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 888888888 * 1e6 as u64,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 777777777 * 1e6 as u64,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 666666666 * 1e6 as u64,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 555555555 * 1e6 as u64,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2 + 1,
                actual_pay: 34444 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2 - 1,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 444444444 * 1e6 as u64,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 333333333 * 1e6 as u64,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 222222222 * 1e6 as u64,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY * 2 / 10 + 1,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2 - 1,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 444444444 * 1e6 as u64,
                actual_pay: 5333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 333333333 * 1e6 as u64,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 222222222 * 1e6 as u64,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY * 2 / 10 + 1,
                actual_pay: 35333333 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 10e8 as u64 * 1e6 as u64,
                actual_pay: 32767 * LAMPORTS_PER_SOL,
            },
            TestCase {
                remaining_supply: 10e8 as u64 * 1e6 as u64 - 1,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 999999999 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 888888888 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 777777777 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 666666666 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 555555555 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2 + 1,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY / 2 - 1,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 444444444 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 333333333 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: 222222222 * 1e6 as u64,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY * 2 / 10 + 1,
                actual_pay: 33333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY * 2 / 10 + 1,
                actual_pay: 333333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY * 1 / 10 + 1,
                actual_pay: 3333333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY,
                actual_pay: 3333333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY - 1,
                actual_pay: 3333333333 * 333333333,
            },
            TestCase {
                remaining_supply: MAX_TOKEN_SUPPLY - 11111111111,
                actual_pay: 3333333333 * 333333333,
            },
            TestCase {
                remaining_supply: token_math::CURVE_1_PARAMS.token_supply_at_boundary - 11111111111,
                actual_pay: 3333333333 * 333333333,
            },
            TestCase {
                remaining_supply: token_math::CURVE_2_PARAMS.token_supply_at_boundary - 11111111111,
                actual_pay: 3333333333 * 333333333,
            },
        ];
        for case in cases.iter() {
            let buy_amount = compute_buy_token_exact_in(case.actual_pay, case.remaining_supply).unwrap();
            let expect_pay = compute_swap(buy_amount, case.remaining_supply, true).unwrap();
            assert!(expect_pay <= case.actual_pay);
        }
    }

    #[test]
    fn test_compute_buy_token_exact_in_with_u64_max() {
        let buy_amount = compute_buy_token_exact_in(u64::MAX, MAX_TOKEN_SUPPLY).unwrap();
        println!("buy_amount: {}", buy_amount);
    }

    #[test]
    fn test_compute_buy_token_exact_in_2() {
        let buy_amount = compute_buy_token_exact_in(100000000, 997000291850416).unwrap();
        println!("buy_amount: {}", buy_amount);
        let pay = compute_swap(buy_amount, 997000291850416, true).unwrap();
        println!("pay: {}", pay);
    }

    #[test]
    fn test_compute_fee() {
        assert_eq!(compute_fee(10000), 100);
        assert_eq!(compute_fee(33), 1);
        assert_eq!(compute_fee(333), 4);
    }

    #[test]
    fn test_split_pay_amount() {
        let (pay, fee) = split_pay_amount(10000).unwrap();
        assert_eq!(pay, 9901);
        assert_eq!(fee, 99);

        let (pay, fee) = split_pay_amount(33).unwrap();
        assert_eq!(pay, 33);
        assert_eq!(fee, 0);

        let (pay, fee) = split_pay_amount(333).unwrap();
        assert_eq!(pay, 330);
        assert_eq!(fee, 3);
    }
}
